### PR TITLE
Loki: Add the ability to prettify logql queries

### LINKED ISSUE #61041: Loki Query Editor: prettify query
In https://github.com/grafana/loki/pull/7906 an andpoint was implemented to prettify queries. This is an interesting and potentially useful feature, which we could support, for example, by adding a dedicated button or other defined behavior.

It's simple to implement, and could help to optimize workflows or reduce cognitive noise in complex/long queries.

### PR DESCRIPTION
**Pull request by itsgareth**
**What is this feature?**
adds a button giving the user an option to prettify their loki query.

**Which issue(s) does this PR fix?**:
Fixes #61041

**how to test this pr**:

**in grafana/grafana**
1. checkout to this branch
3. make sure you have enabled the feature flag `lokiFormatQuery`

### CODE REVIEW

File: public/app/plugins/datasource/loki/components/monaco-query-field/MonacoQueryField.tsx [hunk 1]
```diff
@@ -1,12 +1,12 @@
 import { css } from '@emotion/css';
```
Comment (matyax) [inline public/app/plugins/datasource/loki/components/monaco-query-field/MonacoQueryField.tsx:1]: My first request would be to move the button and all the extra formatting code to `LokiQueryCodeEditor`. This is a component that is used in other places, and it doesn't need to know about the formatting feature, it's a Code Editor feature.

Feel free to pass down an *optional*  `value` prop.

File: public/app/plugins/datasource/loki/datasource.ts [hunk 2]
```diff
@@ -451,6 +451,31 @@ export class LokiDatasource
     return statsForAll;
   }
 
+  async formatQuery(query: string) {
+    const placeHolderScopedVars: any = {
```
Comment (matyax) [inline public/app/plugins/datasource/loki/datasource.ts:455]: Just a reminder that you can (and should) import the ones from `validation.ts`.

File: public/app/plugins/datasource/loki/datasource.ts [hunk 3]
```diff
@@ -451,6 +451,31 @@ export class LokiDatasource
...
+    const interpolated = this.interpolateString(query, placeHolderScopedVars);
+    let response = await this.metadataRequest('format_query', { query: interpolated });
+
+    for (const variable in placeHolderScopedVars) {
+      const value = placeHolderScopedVars[variable].value;
+      response = response.replace(new RegExp(`\\[${value}\\]`, 'g'), `[$${variable}]`);
+    }
+
+    return response;
+  }
```
Comment (matyax) [inline public/app/plugins/datasource/loki/datasource.ts:477]: For now, it will never be trivial to replace the interpolated value back to the original variable name without the risk of  replacing part of the original query by mistake.

For example, I have this variable called `$var` whose value is `2`:



Imagine I want to replace after formatting `2` with `$var`, which `2` of the formatted query are we going to replace? We don't know.

I will try to come up with an alternative suggestion _leveraging_ Lezer to replace the values and not using simple string searches.

In the mean time, I would recommend you to remove the interpolation part, and move this code out of the datasource class to avoid adding more complexity to it, like we did in `querySplitting.ts`. It would also make testing easier.

**Review by matyax — COMMENTED** [review]
Good work so far! I will try to come back to this with a potential approach using Lezer to replace the template variables.

**Comment by matyax** [comment]
Leaving this as a comment to continue later. I've been playing with using the paser to get the node that points to a variable, to be used after formatting to replace the interpolated string with the variable name. For example, I would like to keep a reference to the `Range` node in this tree:

imagen

So I can use it later to replace it with the variable name using the parsed prettified query:

imagen

To attempt to do that, I would store a path to the node (node parents until the root of the tree), for every detected variable.

Datasource:

```javascript
  async formatQuery(query: string) {
    const transformedQuery = transformForFormatting(query, this.interpolateString.bind(this));
    
    let formatted = await this.metadataRequest('format_query', { query: transformedQuery.query  });

    return formatted;
  }
```

Formatter.ts

```javascript
import { SyntaxNode } from '@lezer/common';

import { ScopedVars, TypedVariableModel } from '@grafana/data';
import { parser } from '@grafana/lezer-logql';
import { ErrorId } from 'app/plugins/datasource/prometheus/querybuilder/shared/parsingUtils';

import { placeHolderScopedVars } from './components/monaco-query-field/monaco-completion-provider/validation';

type Transformation = {
    original: string;
    replaced: string;
}

type TransformedQuery = {
    transformations: Transformation[];
    query: string;
}

interface ParseError {
    text: string;
    node: SyntaxNode;
}



function parseQuery(query: string) {
    const parseErrors = new Map<SyntaxNode | null, ParseError[]>();
    const tree = parser.parse(query);
    tree.iterate({
      enter: (nodeRef): false | void => {
        if (nodeRef.type.id === ErrorId) {
          const node = nodeRef.node;
          const siblings = parseErrors.get(node.parent) || []; 
          parseErrors.set(node.parent, [...siblings, {
            node,
            text: query.substring(node.from, node.to),
          }]);
        }
      },
    });
    return parseErrors;
  }

export function transformForFormatting(query: string, interpolateString: (string: string, scopedVars?: ScopedVars) => string): TransformedQuery {
    // We get the possible variables from the parse errors
    const errors = parseQuery(query);
    // Nothing to fix, the query is ready for formatting
    if (!errors.size) {
        return {
            transformations: [],
            query,
        };
    }

    errors.forEach((parseError, parent) => {
        if (parseError[0].text === '$') {
            const variable = query.substring(parseError[0].node.from, parseError[parseError.length-1].node.from);
            const chain = [parent?.type.name];
            let lastParent = parent;
            while (lastParent?.parent) {
                chain.push(lastParent.parent.type.name);
                lastParent = lastParent.parent;
            }
            console.log(variable, interpolateString(variable, placeHolderScopedVars), chain);
        }
    });

    return {
        transformations: [],
        query,
    };
}

export function revertTransformations(query: string, transformations: Transformation[]): string {
    return '';
}
```

Anyway, given the complexity that this could require, I wonder if we should either:
- Ask the backend to be nice with variables somehow.
- Format in the frontend and handle variables better than the endpoint.

In any case, this could be useful to you as an introduction to Lezer. To be continued.

**Comment by matyax** [comment]
We might try something like this: https://github.com/grafana/grafana/pull/64591 Let me know what you think.

File: pkg/tsdb/loki/loki.go [hunk 4]
```diff
@@ -124,8 +125,14 @@ func callResource(ctx context.Context, req *backend.CallResourceRequest, sender
...
 
+	contentType := rawLokiResponse.ContentType
+
+	if contentType == "" {
+		contentType = "application/json"
+	}
+
 	respHeaders := map[string][]string{
-		"content-type": {"application/json"},
+		"content-type": {contentType},
```
Comment (svennergr) [inline pkg/tsdb/loki/loki.go:135]: I guess this is needed because the formatting API returns `text/plain` or something like that, but I don't really like this. This might be far fetched, but with this we are turning our datasource in a "raw" proxy, which would easily allow things like XSS to happen.

File: public/app/plugins/datasource/loki/components/LokiQueryEditor.tsx [hunk 5]
```diff
@@ -91,6 +91,10 @@ export const LokiQueryEditor = React.memo<LokiQueryEditorProps>((props) => {
     setLabelBrowserVisible((visible) => !visible);
   };
 
+  const onPrettifyQuery = async () => {
+    onChangeInternal({ ...query, expr: await datasource.formatQuery(query.expr) });
```
Comment (svennergr) [inline public/app/plugins/datasource/loki/components/LokiQueryEditor.tsx:95]: Hm, why are we using `onChangeInternal` and not just `onChange`?

File: public/app/plugins/datasource/loki/formatter.ts
Comment (svennergr) [inline public/app/plugins/datasource/loki/formatter.ts:1]: I am not a big fan of using `ParseErrors` to determine positions of variables. What happened to the alternative solutions for the prettifying approach?

**Review by svennergr — DISMISSED** [review]

File: public/app/plugins/datasource/loki/formatter.ts
Comment (ivanahuckova) [reply public/app/plugins/datasource/loki/formatter.ts:1]: > What happened to the alternative solutions for the prettifying approach?

+1 on this question

Comment (gwdawson) [reply public/app/plugins/datasource/loki/formatter.ts:1]: this was originally a pr to test the waters and see if this could be a quick win by using the API. @matyax then made a pr with an improved solution to my original implementation using lezer and i just decided to continue from there as i liked this new solution.

i'm still open to discuss the other solutions if this is not something you would like to go ahead with.

File: pkg/tsdb/loki/loki.go (hunk 4 above)
Comment (svennergr) [reply pkg/tsdb/loki/loki.go]: Btw, Prometheus also returns JSON on their `/format_query` responses: https://prometheus.io/docs/prometheus/latest/querying/api/#formatting-query-expressions

Comment (gwdawson) [reply pkg/tsdb/loki/loki.go]: @svennergr this could be a topic to discuss in thursdays sync. try to find out why **text/plain** was decided on and if it could possibly be changed for consistency.

File: lerna.json [hunk 6]
```diff
@@ -1,8 +1,6 @@
 {
   "npmClient": "yarn",
   "useWorkspaces": true,
-  "packages": [
-    "packages/*"
-  ],
+  "packages": ["packages/*"],
```
Comment (matyax) [inline lerna.json:4]: You should probably remove this change.

File: public/app/plugins/datasource/loki/formatterTS.ts [hunk 7]
```diff
@@ -0,0 +1,406 @@
...
+    }
+    if (labelValueA > labelValueB) {
+      return 1;
+    }
+    return 0;
+  });
+
+  // add the formatted label to the response
+  labelNodes.forEach((node) => {
+    const labelNode = node.getChild(Identifier)!;
```
Comment (matyax) [inline public/app/plugins/datasource/loki/formatterTS.ts:102]: You should not use the non-null assertion at all, or as little as possible. There is a reason why TS wants you to check that `labelNode` is not actually null, and it's because it _can_ be null. A malformed query will cause all sort of errors and unexpected running conditions in your code, so even if it's annoying, the whole purpose of TS is to protect you, the code, and the user from these kind of situations.

imagen

File: public/app/plugins/datasource/loki/formatterTS.ts [hunk 8]
```diff
@@ -0,0 +1,406 @@
...
+  Eql,
+  Duration,
+  Bytes,
+  Number,
+  LabelFormatMatcher,
+  FilterOp,
+} from '@grafana/lezer-logql';
+
+export const formatLogQL = (query: string): string => {
+  const tree = parser.parse(query);
```
Comment (matyax) [inline public/app/plugins/datasource/loki/formatterTS.ts:45]: To save you headaches, you should run the query through `validateQuery`, and avoid validation if the query is invalid. That would simplify the implementation *a lot*, and prevent unexpected errors.

When that happens, we can display a message to the user, or disable the formatting button.

File: public/app/plugins/datasource/loki/formatterTS.ts [hunk 9]
```diff
@@ -0,0 +1,406 @@
...
+
+    if (node.getChild(Eq)) {
+      operatorNode = node.getChild(Eq)!;
+    } else if (node.getChild(Neq)) {
+      operatorNode = node.getChild(Neq)!;
+    } else if (node.getChild(Re)) {
+      operatorNode = node.getChild(Re)!;
+    } else {
+      operatorNode = node.getChild(Nre)!;
+    }
```
Comment (matyax) [inline public/app/plugins/datasource/loki/formatterTS.ts:114]: Since we know the query is valid from line 45, we can simplify this to just:

```suggestion
    operatorNode = labelNode.nextSibling;
```

File: public/app/plugins/datasource/loki/formatterTS.ts [hunk 10]
```diff
@@ -0,0 +1,406 @@
...
+export function formatLabelFilter(node: SyntaxNode, query: string): string {
+  if (node.getChild(Matcher)) {
+    const matcherNode = node.getChild(Matcher)!;
+
+    const identifierNode = matcherNode.getChild(Identifier)!;
+    const valueNode = matcherNode.getChild(String)!;
+    const identifier = query.substring(identifierNode.from, identifierNode.to);
+    const value = query.substring(valueNode.from, valueNode.to);
+
+    if (matcherNode.getChild(Eq)) {
```
Comment (matyax) [inline public/app/plugins/datasource/loki/formatterTS.ts:239]: Here you could also use something like `nextSibling` and avoid the literal check.

File: public/app/plugins/datasource/loki/formatterTS.ts [hunk 11]
```diff
@@ -0,0 +1,406 @@
...
+  if (node.getChild(IpLabelFilter)) {
+    const ipLabelFilterNode = node.getChild(IpLabelFilter)!;
+
+    const identifierNode = ipLabelFilterNode.getChild(Identifier)!;
+    const valueNode = ipLabelFilterNode.getChild(String)!;
+    const identifier = query.substring(identifierNode.from, identifierNode.to);
+    const value = query.substring(valueNode.from, valueNode.to);
+
+    if (ipLabelFilterNode.getChild(Eq)) {
+      return `| ${identifier}=ip(${value})`;
```
Comment (matyax) [inline public/app/plugins/datasource/loki/formatterTS.ts:259]: Here you could also use something like `nextSibling` and avoid the literal check.

File: public/app/plugins/datasource/loki/formatterTS.ts [hunk 12]
```diff
@@ -0,0 +1,406 @@
...
+  if (node.getChild(UnitFilter) && node.getChild(UnitFilter)!.getChild(DurationFilter)) {
+    const subNode = node.getChild(UnitFilter)!;
+    const durationFilterNode = subNode.getChild(DurationFilter)!;
+
+    const identifierNode = durationFilterNode.getChild(Identifier)!;
+    const valueNode = durationFilterNode.getChild(Duration)!;
+    const identifier = query.substring(identifierNode.from, identifierNode.to);
+    const value = query.substring(valueNode.from, valueNode.to);
+
+    if (durationFilterNode.getChild(Gtr)) {
```
Comment (matyax) [inline public/app/plugins/datasource/loki/formatterTS.ts:274]: Here you could also use something like `nextSibling` and avoid the literal check.
(Apologies for the repeated comments, but don't want you to miss this).

File: public/app/plugins/datasource/loki/formatterTS.ts [hunk 13]
```diff
@@ -0,0 +1,406 @@
...
+  if (node.getChild(UnitFilter) && node.getChild(UnitFilter)!.getChild(BytesFilter)) {
+    const subNode = node.getChild(UnitFilter)!;
+    const bytesFilterNode = subNode.getChild(BytesFilter)!;
+
+    const identifierNode = bytesFilterNode.getChild(Identifier)!;
+    const valueNode = bytesFilterNode.getChild(Bytes)!;
+    const identifier = query.substring(identifierNode.from, identifierNode.to);
+    const value = query.substring(valueNode.from, valueNode.to);
+
+    if (bytesFilterNode.getChild(Gtr)) {
```
Comment (matyax) [inline public/app/plugins/datasource/loki/formatterTS.ts:300]: Here you could also use something like `nextSibling` and avoid the literal check.

File: public/app/plugins/datasource/loki/formatterTS.ts [hunk 14]
```diff
@@ -0,0 +1,406 @@
...
+
+  if (node.getChild(NumberFilter)) {
+    const numberFilterNode = node.getChild(NumberFilter)!;
+
+    const identifierNode = numberFilterNode.getChild(Identifier)!;
+    const valueNode = numberFilterNode.getChild(Number)!;
+    const identifier = query.substring(identifierNode.from, identifierNode.to);
+    const value = query.substring(valueNode.from, valueNode.to);
+
+    if (numberFilterNode.getChild(Gtr)) {
```
Comment (matyax) [inline public/app/plugins/datasource/loki/formatterTS.ts:325]: Here you could also use something like `nextSibling` and avoid the literal check.

**Review by matyax — COMMENTED** [review]
Reviewed and tested the TS formatter and it looks really really good! Left some coments and suggestions. Amazing work.

**Comment by matyax** [comment]
@gwdawson Found the bug that is currently duplicating the output of `count(sum(rate({compose_project="tns-custom"}[1s])))`.

The root cause of the problem is the function `iterateNode`. What happens is that in `formatVectorAggregationExpr` you call `iterateNode` asking for the following nodes: `[VectorOp, Number, MetricExpr, Grouping]`, and because it's recursive (and unfortunately it's bugged), it returns a lot more that you're expecting.

Instead of getting an array of `VectorOp, MetricExpr` from:

imagen

You get 2 times those two:

imagen

All ☝️  are recursive calls over the same nodes. So, you correctly format the first two `VectorOp, MetricExpr`, and when you're done, you append the extra `VectorOp, MetricExpr`.

From that, the following change fixes the bug:

```javascript
function iterateNode(node: SyntaxNode, lookingFor: number[]): SyntaxNode[] {
  iterateNode2(node, lookingFor);
  const nodes: SyntaxNode[] = [];
  let child = node.firstChild;

  while (child) {
    if (lookingFor.includes(child.type.id)) {
      nodes.push(child);
      lookingFor.splice(lookingFor.indexOf(child.type.id), 1); // this
    }

    nodes.push(...iterateNode(child, lookingFor));
    child = child.nextSibling;
  }

  return nodes;
}
```

imagen

The problem with that change is that it breaks other code that is dependent on `iterateNode` looking for repeated nodes.

One of the main issues with `iterateNode` is that it should be either recursive or use a loop, not both, and I'm not sure why you added a recursive call.

For example:

```javascript
function iterateNode(node: SyntaxNode, lookingFor: number[]): SyntaxNode[] {
  iterateNode2(node, lookingFor);
  const nodes: SyntaxNode[] = [];
  let child = node.firstChild;

  while (child) {
    if (lookingFor.includes(child.type.id)) {
      nodes.push(child);
    }

    // nodes.push(...iterateNode(child, lookingFor)); This should not be necessary
    child = child.nextSibling; // maybe this while should move to the children if there are no more siblings?
  }

  return nodes;
}
```

### TL;DR

Adding `lookingFor.splice(lookingFor.indexOf(child.type.id), 1);` fixes the bug for that query, but the root cause of this bug (and maybe other bugs we don't know about yet) is that `iterateNode` is a buggy implementation that should be fixed.

### Other things I noticed

I would suggest that you avoid recursive calls, specially calls to `formatLokiQuery`, because it's a potential source for bugs. E.g. https://github.com/grafana/grafana/blob/610c4209d2d4041b45df3ba43568c6d384c70052/public/app/plugins/datasource/loki/formatter.ts#L329-L331

There, I would change it to:

```javascript
const metricExpr = query.substring(node.from, node.to); // slice the MetricExpr part
const metricNode = getNodeFromQuery(metricExpr, MetricExpr); // Get the new MetricExpr node
if (metricNode) {
  response += indentMultiline(formatMetricExpr(metricNode, metricExpr), 1); // Pass it to formatMetricExpr
}
```

File: public/app/plugins/datasource/loki/queryUtils.ts [hunk 15]
```diff
@@ -293,3 +297,34 @@ export const getLokiQueryFromDataQuery = (query?: DataQuery): LokiQuery | undefi
 
   return query;
 };
+
+export function formatLogqlQuery(query: string, datasource: LokiDatasource) {
+  reportInteraction('grafana_loki_format_query_clicked', {});
```
Comment (matyax) [inline public/app/plugins/datasource/loki/queryUtils.ts:302]: I think we could also add some extra formation here like if it's a logs query, metric query, query with errors.

Comment (matyax) [reply public/app/plugins/datasource/loki/queryUtils.ts:302]: For example, if they press it when they have errors often, which we don't support, we might do something about it, e.g. messages, etc.

File: public/app/plugins/datasource/loki/queryUtils.ts [hunk 16]
```diff
@@ -293,3 +297,39 @@ export const getLokiQueryFromDataQuery = (query?: DataQuery): LokiQuery | undefi
 
   return query;
 };
+
+export function formatLogqlQuery(query: string, datasource: LokiDatasource) {
```
Comment (matyax) [inline public/app/plugins/datasource/loki/queryUtils.ts:301]: Before merging, we need a couple of test cases here, specially for the transformation part which deals with variables.

**Review by matyax — COMMENTED** [review]
One last thing and we're done here 🚢

**Review by matyax — APPROVED** [review]
Very nice work 👏 👏 👏
